File: android/app/src/main/java/com/limebooking/printbridge/printer/ImageRasterizer.kt
package com.limebooking.printbridge.printer

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.util.Base64

/**
 * Converts a base64 PNG (typically a QR code from Lime's `additionalReceiptData.QR`,
 * or a header logo) into ESC/POS raster bitmap commands using the GS v 0 sequence,
 * which all common thermal printers support.
 *
 * Format reference: ESC/POS GS v 0 m xL xH yL yH d1...dk
 *   m  = density (0 = single, 1 = double width, 2 = double height, 3 = quad)
 *   xL/xH = bitmap width in BYTES (8 dots per byte)
 *   yL/yH = bitmap height in DOTS
 */
object ImageRasterizer {

    fun decodeBase64(base64: String): Bitmap? {
        return try {
            val cleaned = base64.substringAfter("base64,") // strip data: prefix if present
            val bytes = Base64.decode(cleaned, Base64.DEFAULT)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        } catch (e: Exception) {
            null
        }
    }

    private fun scale(src: Bitmap, maxWidth: Int): Bitmap {
        if (src.width <= maxWidth) return src
        val ratio = maxWidth.toFloat() / src.width
        val newH = (src.height * ratio).toInt()
        return Bitmap.createScaledBitmap(src, maxWidth, newH, true)
    }

    private fun toMonochrome(src: Bitmap): BooleanArray {
        // true = black (print), false = white (skip)
        val pixels = IntArray(src.width * src.height)
        src.getPixels(pixels, 0, src.width, 0, 0, src.width, src.height)
        val out = BooleanArray(pixels.size)
        for (i in pixels.indices) {
            val c = pixels[i]
            val a = Color.alpha(c)
            if (a < 128) { out[i] = false; continue }
            val luma = (Color.red(c) * 0.299 + Color.green(c) * 0.587 + Color.blue(c) * 0.114).toInt()
            out[i] = luma < 128
        }
        return out
    }

    /**
     * Scales [bitmap] so width <= [maxWidthDots], converts to 1-bit, and emits
     * the ESC/POS GS v 0 raster command stream.
     */
    fun toEscPosBytes(bitmap: Bitmap, maxWidthDots: Int): ByteArray {
        val scaled = scale(bitmap, maxWidthDots)
        val w = scaled.width
        val h = scaled.height
        val mono = toMonochrome(scaled)

        // Round width up to byte boundary
        val widthBytes = (w + 7) / 8
        val data = ByteArray(widthBytes * h)
        for (y in 0 until h) {
            for (x in 0 until w) {
                if (mono[y * w + x]) {
                    val byteIdx = y * widthBytes + (x / 8)
                    data[byteIdx] = (data[byteIdx].toInt() or (0x80 shr (x % 8))).toByte()
                }
            }
        }

        // GS v 0 m xL xH yL yH data...
        val header = byteArrayOf(
            0x1D, 0x76, 0x30, 0x00,
            (widthBytes and 0xFF).toByte(),
            ((widthBytes shr 8) and 0xFF).toByte(),
            (h and 0xFF).toByte(),
            ((h shr 8) and 0xFF).toByte()
        )
        return header + data
    }
}

File: ios/LimePrintBridge/Printer/ImageRasterizer.swift
import CoreGraphics
import Foundation
import UIKit

/// Converts a base64 PNG (typically a QR code from Lime's
/// `additionalReceiptData.QR`, or a header logo) into ESC/POS raster bitmap
/// commands using the `GS v 0` sequence, supported by all common thermal
/// printers.
///
/// Format: `GS v 0 m xL xH yL yH d1...dk`
///   - m = density (0 = single)
///   - xL/xH = bitmap width in BYTES (8 dots per byte)
///   - yL/yH = bitmap height in DOTS
enum ImageRasterizer {

    static func decodeBase64(_ b64: String) -> UIImage? {
        // Strip any "data:image/...;base64," prefix the PWA might include
        let cleaned = b64.contains(",") ? String(b64.split(separator: ",").last ?? "") : b64
        guard let data = Data(base64Encoded: cleaned, options: .ignoreUnknownCharacters) else { return nil }
        return UIImage(data: data)
    }

    static func toEscPos(_ image: UIImage, maxWidthDots: Int) -> Data {
        guard let cg = image.cgImage else { return Data() }
        let scaled = scale(cg, maxWidth: maxWidthDots)
        let mono = monochromeBytes(scaled)
        return rasterCommand(width: scaled.width, height: scaled.height, bitmap: mono)
    }

    private static func scale(_ image: CGImage, maxWidth: Int) -> CGImage {
        if image.width <= maxWidth { return image }
        let ratio = Double(maxWidth) / Double(image.width)
        let newH = max(1, Int(Double(image.height) * ratio))
        let space = CGColorSpaceCreateDeviceRGB()
        guard let ctx = CGContext(
            data: nil, width: maxWidth, height: newH,
            bitsPerComponent: 8, bytesPerRow: 0, space: space,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return image }
        ctx.interpolationQuality = .high
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: maxWidth, height: newH))
        return ctx.makeImage() ?? image
    }

    /// Returns the bitmap packed as ESC/POS expects: row-major, MSB-first per byte,
    /// 1 = black/print, 0 = white/skip.
    private static func monochromeBytes(_ image: CGImage) -> Data {
        let w = image.width
        let h = image.height
        let bytesPerPixel = 4
        let bytesPerRow = w * bytesPerPixel
        var pixels = [UInt8](repeating: 0, count: bytesPerRow * h)
        let space = CGColorSpaceCreateDeviceRGB()
        guard let ctx = CGContext(
            data: &pixels, width: w, height: h,
            bitsPerComponent: 8, bytesPerRow: bytesPerRow, space: space,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
                | CGBitmapInfo.byteOrder32Big.rawValue
        ) else { return Data() }
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: w, height: h))

        let widthBytes = (w + 7) / 8
        var out = Data(count: widthBytes * h)
        out.withUnsafeMutableBytes { buf -> Void in
            let bytes = buf.bindMemory(to: UInt8.self).baseAddress!
            for y in 0..<h {
                for x in 0..<w {
                    let off = y * bytesPerRow + x * bytesPerPixel
                    let r = Int(pixels[off])
                    let g = Int(pixels[off + 1])
                    let b = Int(pixels[off + 2])
                    let a = Int(pixels[off + 3])
                    let isBlack: Bool
                    if a < 128 {
                        isBlack = false
                    } else {
                        let luma = (r * 299 + g * 587 + b * 114) / 1000
                        isBlack = luma < 128
                    }
                    if isBlack {
                        let byteIdx = y * widthBytes + (x / 8)
                        let bit: UInt8 = 0x80 >> (x % 8)
                        bytes[byteIdx] |= bit
                    }
                }
            }
        }
        return out
    }

    private static func rasterCommand(width: Int, height: Int, bitmap: Data) -> Data {
        let widthBytes = (width + 7) / 8
        var out = Data()
        // GS v 0 m xL xH yL yH
        out.append(contentsOf: [
            0x1D, 0x76, 0x30, 0x00,
            UInt8(widthBytes & 0xFF),
            UInt8((widthBytes >> 8) & 0xFF),
            UInt8(height & 0xFF),
            UInt8((height >> 8) & 0xFF)
        ])
        out.append(bitmap)
        out.append(contentsOf: [0x0A]) // line feed
        return out
    }
}
